### DISCUSSION TITLE
uart_tx and uart_rx_enable can not work and returns "-19"

### DISCUSSION THREAD

**Discussion by ZiminCN** [zephyrproject-rtos/zephyr#57975] [Q&A]
hi! I have been using zephyr recently， and i found some issues while debugging the serial port (UART)，I attempt to call the asynchronous serial port in the following order. However, the serial port does not have any data output and cannot accept data.
and uart_tx and uart_rx_enable both returns "-19", I don't understand what this return value represents, how should i solve it?

for more detailed information, i am using STM32G070KBT6 and the zephyr version is 3.3.0, I have already set the CONFIG_UART_CONSOLE=y
CONFIG_UART_ASYNC_API=y
CONFIG_UART_USE_RUNTIME_CONFIGURE=y
at the MyPrj_defconfig

Here is my rough code:

/* code start */

static uint8_t rx_buf[10] = {0};
uint8_t another_rx_buf[10] = {0};
static uint8_t tx_buf[] = {"**Test Zephyr Uart Transmit Func Success!** \n\r"};

void uartInterrupt(const struct device *uart, struct uart_event *evt, void *user_data){

	switch (evt->type) {
	case UART_TX_DONE:

		break;
	case UART_TX_ABORTED:

		break;
	case UART_RX_RDY:
		uart_rx_enable(uart, rx_buf, sizeof(rx_buf), 100);
		break;
	case UART_RX_BUF_REQUEST:
		uart_rx_buf_rsp(uart, another_rx_buf, sizeof(another_rx_buf));
		break;
	case UART_RX_BUF_RELEASED:

		break;
	case UART_RX_DISABLED:
		uart_rx_enable(uart, rx_buf, sizeof(rx_buf), 100);
		break;
	case UART_RX_STOPPED:

		break;

	default:
		break;
	}
}

int main(void)
{	
	uart_config my_config;
	my_config.baudrate = 460800;
	my_config.parity = UART_CFG_PARITY_NONE;
	my_config.stop_bits = UART_CFG_STOP_BITS_1;
	my_config.data_bits = UART_CFG_DATA_BITS_8;
	my_config.flow_ctrl = UART_CFG_FLOW_CTRL_NONE;

	volatile int err_configure = 1;
	volatile int err_set = 1;
	volatile int err_rx_enable = 1;
	volatile int err_tx = 1;

	err_configure = uart_configure(head_uart, &my_config);
	err_set = uart_callback_set(head_uart, uartInterrupt, NULL);
	err_rx_enable = uart_rx_enable(head_uart, rx_buf, sizeof(rx_buf), 100);
	err_tx = uart_tx(head_uart, tx_buf, sizeof(tx_buf), 1000);

	while (1)
	{
		k_usleep(100000);
	}
	return 0;
}
/* code end */

**Comment by ArielElb** [comment]
Did you found any solution? 
Thx

> hi! I have been using zephyr recently， and i found some issues while debugging the serial port (UART)，I attempt to call the asynchronous serial port in the following order. However, the serial port does not have any data output and cannot accept data.
> and uart_tx and uart_rx_enable both returns "-19", I don't understand what this return value represents, how should i solve it?
> 
> for more detailed information, i am using STM32G070KBT6 and the zephyr version is 3.3.0, I have already set the CONFIG_UART_CONSOLE=y
> CONFIG_UART_ASYNC_API=y
> CONFIG_UART_USE_RUNTIME_CONFIGURE=y
> at the MyPrj_defconfig
> 
> Here is my rough code:
> 
> /* code start */
> 
> static uint8_t rx_buf[10] = {0};
> uint8_t another_rx_buf[10] = {0};
> static uint8_t tx_buf[] = {"**Test Zephyr Uart Transmit Func Success!** \n\r"};
> 
> void uartInterrupt(const struct device *uart, struct uart_event *evt, void *user_data){
> 
> 	switch (evt->type) {
> 	case UART_TX_DONE:
> 
> 		break;
> 	case UART_TX_ABORTED:
> 
> 		break;
> 	case UART_RX_RDY:
> 		uart_rx_enable(uart, rx_buf, sizeof(rx_buf), 100);
> 		break;
> 	case UART_RX_BUF_REQUEST:
> 		uart_rx_buf_rsp(uart, another_rx_buf, sizeof(another_rx_buf));
> 		break;
> 	case UART_RX_BUF_RELEASED:
> 
> 		break;
> 	case UART_RX_DISABLED:
> 		uart_rx_enable(uart, rx_buf, sizeof(rx_buf), 100);
> 		break;
> 	case UART_RX_STOPPED:
> 
> 		break;
> 
> 	default:
> 		break;
> 	}
> }
> 
> int main(void)
> {	
> 	uart_config my_config;
> 	my_config.baudrate = 460800;
> 	my_config.parity = UART_CFG_PARITY_NONE;
> 	my_config.stop_bits = UART_CFG_STOP_BITS_1;
> 	my_config.data_bits = UART_CFG_DATA_BITS_8;
> 	my_config.flow_ctrl = UART_CFG_FLOW_CTRL_NONE;
> 
> 	volatile int err_configure = 1;
> 	volatile int err_set = 1;
> 	volatile int err_rx_enable = 1;
> 	volatile int err_tx = 1;
> 
> 	err_configure = uart_configure(head_uart, &my_config);
> 	err_set = uart_callback_set(head_uart, uartInterrupt, NULL);
> 	err_rx_enable = uart_rx_enable(head_uart, rx_buf, sizeof(rx_buf), 100);
> 	err_tx = uart_tx(head_uart, tx_buf, sizeof(tx_buf), 1000);
> 
> 	while (1)
> 	{
> 		k_usleep(100000);
> 	}
> 	return 0;
> }
> /* code end */

**Comment by ldomaigne** [comment]
Hi @ArielElb, 

This discussion thread might help: https://github.com/zephyrproject-rtos/zephyr/discussions/90108

Cheers! Loïc
